src/Middleware/RateLimiting/src/RateLimitingMetrics.cs

```diff
@@ -0,0 +1,161 @@
+
+using System.Diagnostics;
+using System.Diagnostics.Metrics;
+using System.Runtime.CompilerServices;
+using Microsoft.Extensions.Metrics;
+
+namespace Microsoft.AspNetCore.RateLimiting;
+
+internal sealed class RateLimitingMetrics : IDisposable
+{
+    public const string MeterName = "Microsoft.AspNetCore.RateLimiting";
+
+    private readonly Meter _meter;
+    private readonly UpDownCounter<long> _currentLeaseRequestsCounter;
+    private readonly Histogram<double> _leaseRequestDurationCounter;
+    private readonly UpDownCounter<long> _currentRequestsQueuedCounter;
+    private readonly Histogram<double> _queuedRequestDurationCounter;
+    private readonly Counter<long> _leaseFailedRequestsCounter;
+
+    public RateLimitingMetrics(IMeterFactory meterFactory)
+    {
+        _meter = meterFactory.CreateMeter(MeterName);
+
+        _currentLeaseRequestsCounter = _meter.CreateUpDownCounter<long>(
+            "current-leased-requests",
+            description: "Number of HTTP requests that are currently active on the server that hold a rate limiting lease.");
+
+        _leaseRequestDurationCounter = _meter.CreateHistogram<double>(
+            "leased-request-duration",
+            unit: "s",
+            description: "The duration of rate limiting leases held by HTTP requests on the server.");
+
+        _currentRequestsQueuedCounter = _meter.CreateUpDownCounter<long>(
+            "current-queued-requests",
+            description: "Number of HTTP requests that are currently queued, waiting to acquire a rate limiting lease.");
+
+        _queuedRequestDurationCounter = _meter.CreateHistogram<double>(
+            "queued-request-duration",
+            unit: "s",
+            description: "The duration of HTTP requests in a queue, waiting to acquire a rate limiting lease.");
+
+        _leaseFailedRequestsCounter = _meter.CreateCounter<long>(
+            "lease-failed-requests",
+            description: "Number of HTTP requests that failed to acquire a rate limiting lease. Requests could be rejected by global or endpoint rate limiting policies. Or the request could be canceled while waiting for the lease.");
+    }
+
+    public void LeaseFailed(string? policyName, string? method, string? route, RequestRejectionReason reason)
+    {
+        if (_leaseFailedRequestsCounter.Enabled)
+        {
+            LeaseFailedCore(policyName, method, route, reason);
+        }
+    }
+
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    private void LeaseFailedCore(string? policyName, string? method, string? route, RequestRejectionReason reason)
+    {
+        var tags = new TagList();
+        InitializeRateLimitingTags(ref tags, policyName, method, route);
+        tags.Add("reason", reason.ToString());
+        _leaseFailedRequestsCounter.Add(1, tags);
+    }
+
+    public void LeaseStart(string? policyName, string? method, string? route)
+    {
+        if (_currentLeaseRequestsCounter.Enabled)
```

Comment:
Can this be enabled and disabled while the process is running like event counters? If so, if I enable the `_currentLeaseRequestsCounter` when there are a bunch of active requests with active leases, don't we risk undercounting?

I know in KestrelEventSource, we always increment and decrement the backing fields for PollingCounters regardless of whether they are currently enabled. I see that's not the case for KestrelMetrics.

And even though the `_leaseFailedRequestsCounter` can't go negative, we should at least make it clear in the description that this is measuring the failed requests while the counter was enabled, not for the entire duration the application was running like it would be for a typical PollingCounter.

Reply:
> Can this be enabled and disabled while the process is running like event counters? If so, if I enable the _currentLeaseRequestsCounter when there are a bunch of active requests with active leases, don't we risk undercounting?

Yes, good point. I've updated the implementation to cache enable flags for current requests queued and current leased requests. That way if someone starts listening while in the middle of a request, there is no chance of up/down counters going into a negative value.

> And even though the _leaseFailedRequestsCounter can't go negative, we should at least make it clear in the description that this is measuring the failed requests while the counter was enabled, not for the entire duration the application was running like it would be for a typical PollingCounter.

That will be true for every metrics counter. A separate backing field was normal for event counters. Not storing a value in .NET and just passing a delta to the counter is normal for metrics counters. I don't want to add the same information to every counter's description.